app/waf/src/services/bigiq.service.ts

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright 2019 F5 Networks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getService} from '@loopback/service-proxy';
+import {inject, Provider} from '@loopback/core';
+import {BigIqDataSource} from '../datasources';
+import {factory} from '../log4ts';
+import {Logger} from 'typescript-logging';
+import {checkAndWait} from '../utils';
+
+const BIGIQ_HOST: string = process.env.BIGIQ_HOST || 'localhost';
+const BIGIQ_PORT: number = Number(process.env.BIGIQ_PORT) || 8888;
+const BIGIQ_USERNAME: string = process.env.BIGIQ_USERNAME || 'admin';
+const BIGIQ_PASSWORD: string = process.env.BIGIQ_PASSWORD || 'admin';
+const BIGIQ_POOL: string = process.env.BIGIQ_POOL || 'unknown';
+
+type BigIqToken = {
+  token: string;
+  exp: number;
+};
+
+type LoginResponse = {
+  token: BigIqToken;
+};
+
+type RevokeResponse = {
+  id: string;
+};
+
+type QueryResponse = {
+  status: string;
+  errorMessage: string;
+};
+
+export interface BigIqService {
+  login(
+    host: string,
+    port: number,
+    username: string,
+    password: string,
+  ): Promise<LoginResponse>;
+
+  revoke(
+    host: string,
+    port: number,
+    token: string,
+    body: object,
+  ): Promise<RevokeResponse>;
+
+  query(
+    host: string,
+    port: number,
+    token: string,
+    task: string,
+  ): Promise<QueryResponse>;
+}
+
+export class BigIqServiceProvider implements Provider<BigIqService> {
+  constructor(
+    @inject('datasources.bigiq')
+    protected dataSource: BigIqDataSource = new BigIqDataSource(),
+  ) {}
+
+  value(): Promise<BigIqService> {
+    return getService(this.dataSource);
+  }
+}
+
+export class BigIqManager {
+  private service: BigIqService;
+  private logger: Logger;
+  private static token: BigIqToken;
+
+  constructor(svc: BigIqService, reqId = 'Unknown') {
+    this.service = svc;
+    this.logger = factory.getLogger(reqId + ': services.BigIqManager');
+  }
+
+  async login(): Promise<BigIqToken> {
+    let current = new Date().getTime() / 1000;
+    // Login BIG-IQ, if no token or token will expire in 30 seconds
+    if (!BigIqManager.token || current > BigIqManager.token.exp - 30) {
```

Comment:
If the system clock of our control plane is wrong, we will never have a way to calculate the time. There will be two solution: 1. Assume the clock is correct, since it is our control plane. 2. Auto login again, once 401 occurs. We can start a new commit to implement that.

Reply:
I suggest to see how `AuthToken` does.

Reply:
It also requires a correct clock.